src/core/dense_set.cc

```diff
@@ -0,0 +1,424 @@
+      mr()->deallocate(ptr.Raw(), sizeof(DenseLinkKey), alignof(DenseLinkKey));
+    }
+  } else if (ptr.IsLink()) {
+    // if the pointer is already a link then no allocation needed
+    *ptr.Next() = *it;
+    *it = ptr;
+  } else {
+    // allocate a new link if needed and copy the pointer to the new link
+    LinkAllocator la(mr());
+    DenseLinkKey* lk = la.allocate(1);
+    la.construct(lk);
+
+    lk->next = *it;
+    lk->SetObject(ptr.GetObject());
+    it->SetLink(lk);
+  }
+}
+
+auto DenseSet::PopPtrFront(DenseSet::ChainVectorIterator it) -> DensePtr {
+  if (it->IsEmpty()) {
+    return DensePtr(nullptr);
+  }
+
+  DensePtr front = *it;
+
+  // if this is an object, then it's also the only record in this chain.
+  // therefore, we should just reset DensePtr.
+  if (it->IsObject()) {
+    it->Reset();
+  } else {
+    DCHECK(it->IsLink());
+    // this will not return nullptr since it is a Link
+    *it = *it->Next();
+  }
+
+  return front;
+}
+
+void* DenseSet::PopDataFront(DenseSet::ChainVectorIterator it) {
+  DensePtr front = PopPtrFront(it);
+  void* ret = front.GetObject();
+
+  if (front.IsLink()) {
+    DenseLinkKey* lk = (DenseLinkKey*)front.Raw();
+    mr()->deallocate(lk, sizeof(DenseLinkKey), alignof(DenseLinkKey));
+  }
+
+  return ret;
+}
+
+auto DenseSet::Unlink(DenseSet::DensePtr* node) -> DensePtr {
+  DensePtr ret = *node;
+  if (node->IsObject()) {
+    node->Reset();
+  } else {
+    *node = *node->Next();
+  }
+
+  return ret;
+}
+
+void* DenseSet::UnlinkAndFree(DenseSet::DensePtr* node) {
+  DensePtr unlinked = Unlink(node);
+  void* ret = unlinked.GetObject();
+
+  if (unlinked.IsLink()) {
+    mr()->deallocate(unlinked.Raw(), sizeof(DenseLinkKey), alignof(DenseLinkKey));
+  }
+
+  return ret;
+}
+
+void DenseSet::Clear() {
+  for (auto it = entries_.begin(); it != entries_.end(); ++it) {
+    while (!it->IsEmpty()) {
+      PopDataFront(it);
+    }
+  }
+
+  entries_.clear();
+}
+
+auto DenseSet::FindEmptyAround(uint32_t bid) -> ChainVectorIterator {
+  if (entries_[bid].IsEmpty()) {
+    return entries_.begin() + bid;
+  }
+
+  if (!kAllowDisplacements) {
+    return entries_.end();
+  }
+
+  if (bid + 1 < entries_.size() && entries_[bid + 1].IsEmpty()) {
+    return entries_.begin() + bid + 1;
+  }
+
+  if (bid && entries_[bid - 1].IsEmpty()) {
+    return entries_.begin() + bid - 1;
+  }
+
+  return entries_.end();
+}
+
+bool DenseSet::Reserve(size_t sz) {
+  sz = std::min<size_t>(sz, kMinSize);
+
+  if (sz > entries_.max_size()) {
+    return false;
+  }
+
+  sz = absl::bit_ceil(sz);
+  capacity_log_ = absl::bit_width(sz);
+  entries_.reserve(sz);
+
+  if (entries_.capacity() != sz) {
+    return false;
+  }
+
+  return true;
+}
+
+void DenseSet::Grow() {
+  size_t prev_size = entries_.size();
+  entries_.resize(prev_size * 2);
+  ++capacity_log_;
+
+  for (long i = prev_size - 1; i >= 0; --i) {
+    DensePtr* curr = &entries_[i];
+
+    while (curr != nullptr && !curr->IsEmpty()) {
```

Comment:
I would separate these 2 conditions:
1. seems that `curr->IsEmpty()` can happen only for the `&entries_[i]` but never after.
2. `curr is nullptr` happens at the end of the non-empty chain.

so why not use 
```
if (curr->IsEmpty())
  continue;
``` 
before entering the loop and remove it from the loop?

Reply:
The `!curr->IsEmpty()` clause takes place in the loop condition for the case when the list of `entries_[i]` may not start with an empty `DensePtr` but after node removals, becomes empty since every node in the chain has been removed. In this case the root `DensePtr` is reset but the node variable is never moved to the "next" item to become a `nullptr` and exit.

Reply:
maybe add a comment saying that curr can become `empty` but non-null via Unlink call.